idea-plugin/src/main/kotlin/arrow/meta/ide/plugins/proofs/intentions/explicitcoercion.kt

```diff
@@ -1,51 +1,109 @@
 import org.jetbrains.kotlin.psi.KtCallElement
 import org.jetbrains.kotlin.psi.KtElement
+import org.jetbrains.kotlin.psi.KtExpression
 import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.psi.KtPsiFactory
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
 import org.jetbrains.kotlin.types.KotlinType
 
 fun IdeMetaPlugin.makeExplicitCoercionIntention(compilerContext: CompilerContext): ExtensionPhase =
   addIntention(
     text = "Make coercion explicit",
     kClass = KtElement::class.java,
     isApplicableTo = { ktCall: KtElement, _ ->
-      ktCall.explicitParticipatingTypes()?.let { (subtype, supertype) ->
+      //TODO we should handle the caret here to not show the intention for all KtCallElement args,
+      // but only the implicit ones
+      ktCall.explicitParticipatingTypes().any { (subtype, supertype) ->
         compilerContext.areTypesCoerced(subtype, supertype)
-      } ?: false
+      }
     },
     applyTo = { ktCall: KtElement, _ ->
-//      TODO()
+      when (ktCall) {
+        //TODO this should only modify one element at a time
+        is KtCallElement -> {
+          // Get the coerced types from all arguments
+          val targetingTypes: List<PairTypes> = ktCall.explicitParticipatingTypes()
+            .filter { (subtype, supertype) -> compilerContext.areTypesCoerced(subtype, supertype) }
+
+          // get all coerced call element args
+          ktCall.valueArgumentList?.arguments?.mapNotNull { it.getArgumentExpression() }
+            ?.mapNotNull { arg ->
+              val type = arg.resolveType()
+              targetingTypes.firstOrNull { it.subType == type }?.let { targetType ->
+                targetType to arg
+              }
+            }
+            // replace them with the explicit version with the proof
+            ?.forEach { (targetType, arg) ->
+              replaceWithProof(arg, compilerContext, targetType)
+            }
+        }
+
+        is KtProperty -> {
+          ktCall.explicitParticipatingTypes().first().let { pairType ->
+            ktCall.initializer?.let { initializer ->
+              replaceWithProof(initializer, compilerContext, pairType)
+            }
+          }
+        }
+
+      }
     }
   )
 
+private fun KtPsiFactory.replaceWithProof(element: KtExpression, compilerContext: CompilerContext, pairType: PairTypes) {
+  element.replace(
+    createExpression(
+      "${element.text}.${compilerContext.coerceProof(pairType.subType, pairType.superType)?.through?.name}()"
+    )
+  )
+}
+
 //TODO move elsewhere
-fun KtElement.explicitParticipatingTypes(): Pair<KotlinType, KotlinType>? =
+data class PairTypes(val subType: KotlinType, val superType: KotlinType) {
+  companion object {
+    infix fun KotlinType?.pairOrNull(b: KotlinType?): PairTypes? =
+      if (this != null && b != null) PairTypes(this, b)
+      else null
+  }
+}
+
+fun KtElement.explicitParticipatingTypes(): List<PairTypes> =
```

Comment:
there is a function `traverseFilter` with PsiElement as its receiver, which simplifies this block.

Reply:
you mean instead of using this, to call traverseFilter and do the when inside?

Reply:
Nop, traverseFilter works somewhat `similar` to what we have in Arrow. The difference is that it starts at a Node and traverses its whole tree hierarchy matching on nodes you care about, which it will then transform with your desired function `f` resulting with a List<B>. 
This removes most of the blocks collecting and recollecting Lists. Check out the examples in the ide and their call-sites.

Reply:
Let me check, maybe that signature has not been ported to type-proofs. I push something into the type-proofs branch with what I mentioned above.

Reply:
I admit the example in the ide is not really expressing to what terms you can go but as a demo, you may  consider something like:
```
fun KtElement.participatingTypes(): List<KtValueArgument> =
  traverseFilter(KtCallElement::class.java) { call: KtCallElement ->
    call.valueArgumentList?.arguments?.mapNotNull { valueArgs: KtValueArgument ->
      valueArgs
      /*Do something with valueArgs*/
    }
  }.flatten()
```
That algebra is available in the last [commit](https://github.com/arrow-kt/arrow-meta/commit/ae335d763a138075fc5440a6f90fe5bf79211f27)

Reply:
@i-walker The only problem i see on this specific case is that then I'll need to do a traverseFilter per element type we need to support, no? So in this case it'll be

traverseFilter(KtCallElement::class.java) { ... } 
+
traverseFilter(KtProperty::class.java) { ... }

or something like that, no? But in this case I guess we're gonna iterate over all elements and then filter, so I don't see the benefit vs a normal when but I'm sure i'm missing something 🤔

Reply:
let's go for the traverseFilter approach and see what we can abstract over most likely you'll abstract over `f` or choose a type higher in the type hierarchy to simplify the work

Reply:
> you mean instead of using this, to call traverseFilter and do the when inside?

Then either I'm missing something or I come back to this :D 


```kotlin
fun KtElement.explicitParticipatingTypes(): List<PairTypes> =
  traverseFilter(KtElement::class.java) {
    when (it) {
      is KtCallElement -> {
        // Obtain the argument types from the current call
        val subTypes = it.valueArgumentList?.arguments.orEmpty().mapNotNull { it.getArgumentExpression()?.resolveType() }

        val superTypes = analyze(bodyResolveMode = BodyResolveMode.FULL)
          .getSliceContents(BindingContext.RESOLVED_CALL)
          // get the calls for the current element
          .filter { (call, _) -> call.callElement == this }
          // then the argument types
          .entries.first().value.valueArguments
          .map { it.key.type }

        //TODO check for named and switched arguments!

        subTypes.zip(superTypes, ::PairTypes)
      }

      is KtProperty -> {
        listOfNotNull(it.initializer?.resolveType() pairOrNull it.type())
      }

      else -> emptyList()
    }
  }.flatten()
```

Reply:
these are two different functions. Please dont use `when`, when using traverse. That doesn't make sense

Reply:
Thanks @aballano for the meet. I misunderstood the intention of the code :) 
Traverse is not needed

Reply:
Thank you for the time and the explanations! ❤️